docs/rfc/rfc-10.adoc

```diff
@@ -0,0 +1,248 @@
+    // contract depositor can use the extra data to represent additional
+    // information, such as the depositor's address on a foreign (non-Ethereum)
+    // chain.
+    bytes32 depositorExtraData;        
+    // This struct doesn't contain `__gap` property as the structure is not
+    // stored, it is used as a function's calldata argument.
+}
+```
+
+The `DepositRequest` function will gain the same field. Since mapping values are
+stored in different slots we only need to make sure the new field comes after
+the existing ones to address upgradeability concerns.
+
+```
+/// @notice Represents tBTC deposit request data.
+struct DepositRequest {
+    // (...)
+
+    // Optional extra data used by depositors who are smart contracts. The smart
+    // contract depositor can use the extra data to represent additional
+    // information, such as the depositor's address on a foreign (non-Ethereum)
+    // chain.
+    bytes32 depositorExtraData;
+}
+```
+
+To maintain backward compatibility with the existing deposit flow, we will
+implement another reveal function in the `Bridge` contract. The function will be
+mostly identical to the existing `revealDeposit` except that it will respect the
+alternative P2SH deposit script version allowing to pass 32 bytes of extra data.
+The extra data will not be stored in the Bridge. It is the depositor's
+responsibility to make the use of them.
+
+==== Solana gateway
+
+`SolanaGateway` contract will be deployed on Solana and should be placed in
+`tbtc-v2/cross-chain/solana` project. The contract will communicate with
+Ethereum L1 using Wormhole VAAs to reveal Bitcoin deposits. The contract will
+use Wormhole Token Bridge to request Bitcoin redemption on Ethereum L1.
+
+For the implementation, we should consider using the
+link:https://solang.readthedocs.io/en/latest/index.html[Solang compiler]. 
+
+The contract needs to be upgradeable.
+
+==== Solana depositor
+
+`SolanaDepositor` contract will act as tBTC depositor on L1 Ethereum and should
+be placed in `tbtc-v2/solidity` project. The contract will receive Wormhole VAAs
+from Solana and reveal Bitcoin deposits to the tBTC `Bridge` contract. The
+contract will map the revealed deposit to the Solana address. It will also
+bridge the received tBTC once the deposit is swept or tBTC is optimistically
+minted.
+
+This contract needs to have a reference to the `TBTCVault` to inspect the
+`optimisticMintingRequests` mapping and to the `Bridge` to inspect the
+`deposits` mapping to confirm the state when it is notified tBTC has been minted
+for the given deposit. 
+
+The contract needs to be upgradeable.
+
+==== Solana redeemer
+
+`SolanaRedeemer` contract will act as tBTC redeemer on L1 Ethereum and should be
+placed in `tbtc-v2/solidity` project. The contract will receive tBTC from Solana
+via Wormhole Token Bridge and request redemption in the tBTC `Bridge` contract. 
+
+The contract needs to be upgradeable.
+
+==== Relayer bot
```

Comment:
Can we consider adding some guardians bots that will protect against the relayer bot going south?

Reply:
What particular problematic scenario do you have in mind? The relayer should not have any extra rights, it is calling the same functions any anonymous user would call. The only role of the relayer is to call those functions on behalf of the user to improve the UX.